Abstracted Transport

jacksontj opened the pull request:
This is a _massive_ overhaul of the entire transport system within salt (this is the one I mentioned @ SaltConf)

The basic idea in this PR is to abstract out the transport within salt-- since Salt shouldn't care how the bits are shipped to the minion, just that they can and will be. To do this I've taken the "channel" idea and run with it. To do this I've created a "reqchannel" and a "pubchannel" these 2 classes encapsulate the 2 types of messages that salt cares about. For each channel type there are server and client implementations, which means that adding a new transport is simply implementing a base-class and configuring the master.

Now that there is an interface to write transport to I also created a basic TCP implementation which uses the same auth as the current zeromq implementation.

In addition to the abstraction of transport this also introduces tornado as an async programming framework (read: ioloop). This means that we can actually run N transports on a master simultaneously. In addition to making all transport async-- this means we can now make API calls within the master/minion async-- meaning a given MWorker thread could actually have multiple requests in-flight (unlike the single request at a time of today).

In addition to running multiple transports at once this also means we can remove the threading we've been using in MultiMinion and Syndics-- and just use the IOLoop.

There is going to need to be some transition to the new Async interfaces, but this PR includes wrappers which make the async interfaces sync friendly-- so it should be 100% backwards compatible. This just means that if you want your module/runner/etc. to be async (meaning the thread can do something else) you'd have to switch your calls to channels (which are _very_ uncommon) to the async versions.

Before we merge this in I'll need to finish up the docs, but I want to get this opened sooner rather than later to get other people's eyes on it.

jacksontj committed:
Refactor channel classes to have separate pub/req channels

jacksontj committed:
A few hacks to get some basic pubchannel working

jacksontj committed:
Rebase in current develop, and add in recv_noblock

jacksontj committed:
Consolidate socketopts into channel

jacksontj committed:
More work to remove minion.authenticate()

jacksontj committed:
Switch remaining uses of sreq to channel

jacksontj committed:
Misc cleanup after rebase

jacksontj committed:
Move daemon init and execution into the same process.

These daemon processes aren't all thread safe, so having the test suite init then fork is disingenuous since that isn't supported anyways.

jacksontj committed:
Refactor the start_daemon function

jacksontj committed:
Update commented out version as well

jacksontj committed:
Remove some unused imports

jacksontj committed:
Implement timeout for PubChannel.recv()

jacksontj committed:
Cleanup print from debugging

jacksontj committed:
More migration to PubChannel

jacksontj committed:
Add "socket" interface to pub_channel, so we can wire it into regular pollers

jacksontj committed:
Remove all final references to self.socket from minion.py

jacksontj committed:
Refactor minion events to be their own class instead of spread throughout the minion classes

jacksontj committed:
Remove some noe-deprecated and unused functions

jacksontj committed:
Remove old refs to _init_context_and_poller

jacksontj committed:
Pull all fallback cleanup into its own function

jacksontj committed:
Add local channel to factory

jacksontj committed:
Rename channel to client

A bit more clear since there are client and server-side classes

jacksontj committed:
Initial migration of master-side ReqChannels

This is an attempt to pull all transport/crypto logic out of the core master classes.

Once this is completed we should be able to re-use all of this code for all transport types

jacksontj committed:
Initial move of publisher into PubServerChannel

jacksontj committed:
Re-work req_channel in the master daemons, now we don't pass it down the stack, the master-side funcs need to return a dict of opts which specify which send_func to use

jacksontj committed:
Change req_fun default to send (which is encrypted)

jacksontj committed:
Migrate _auth logic into the ReqServerChannel

jacksontj committed:
Misc cleanup

jacksontj committed:
Hacking to fix various things broken in rebase

jacksontj committed:
Initial addition of tcp channel

jacksontj committed:
More hacking to get tcp working

jacksontj committed:
Remove testing code

jacksontj committed:
Cleanup pubchannel polling

zmq poller returns either zmq socket objects of raw FDs. adding an API in the clients to differentiate

jacksontj committed:
Major overhaul of tcp implementation

Now using the same crypto stuff as the zeromq implementation

jacksontj committed:
Re-work PubChannel

- add register_poller (since some people need to change the FD (on failure etc)
- remove recv_noblock

jacksontj committed:
ReqChannel cleanup

- Remove recv_noblock from ReqServer channel
- keepalive
- epoll

jacksontj committed:
Revert register poller, we'll just have the minion exit

jacksontj committed:
Misc cleanup

jacksontj committed:
Whitespace cleanup

jacksontj committed:
Remove rep_hwm (doesn't do anything)

jacksontj committed:
cleanup

jacksontj committed:
Don't make a copy of the list to iterate (it might be very large). Instead keep a list of who to remove

jacksontj committed:
Fix socket recv_frame

jacksontj committed:
Cleanup shutdown of publisher

jacksontj committed:
move publisher to IOLoop

jacksontj committed:
Allow passing in of header

jacksontj committed:
fix typo

jacksontj committed:
Switch ReqServer over to IOLoop

jacksontj committed:
Add un-tested SaltMessageClient

jacksontj committed:
Misc cleanup

jacksontj committed:
Migrate zeromq channel over to ioloop as well

jacksontj committed:
Very basic zmq and tcp support

jacksontj committed:
Switch Req client over to tornado based client

jacksontj committed:
consolidate master-side transport parsing

jacksontj committed:
Logging cleanup

jacksontj committed:
Rough-in of io_loop in minion

jacksontj committed:
Misc cleanup to single minion

jacksontj committed:
Basic multisyndic working now

jacksontj committed:
Cleanup pub channel

Move to on_recv, remove old poller interfaces

jacksontj committed:
cleanup

jacksontj committed:
Add scheduler

jacksontj committed:
Make SaltMessageClient auto-reconnect

jacksontj committed:
Last switch over to iter_transport_opts (support 1->N transports)

jacksontj committed:
Make publisher not wait for writes

jacksontj committed:
loop_interval is a float, lets leave it that way

jacksontj committed:
Don't set 'enc' to clear on clear returns, some jobs don't return dicts

jacksontj committed:
Cleanup some IOLoops

jacksontj committed:
Add timeouts to tcp implementation

jacksontj committed:
Add support for un-registering the on_recv

jacksontj committed:
Cleanup multiminion to work on the ioloop

jacksontj committed:
Remove Event poller (since we are doing IOLoops now)

jacksontj committed:
Since each minion object handles its own events, lets not double send :)

jacksontj committed:
Remove recv() from pub_channels (now we use on_recv)

jacksontj committed:
Remove unecessary imports

jacksontj committed:
Pylint

jacksontj committed:
Add initial req_channel tests

jacksontj committed:
Add basic Pub channel tests

jacksontj committed:
Begin work on async auth/pub/req channels.

For a while we'll have to maintin sync and async versions. The plan is to replace the sync ones with wraps around the async ones.

jacksontj committed:
Handle timeouts in auth/pub connect

jacksontj committed:
Cleanup async

jacksontj committed:
misc cleanup

jacksontj committed:
Misc cleanup to get the minion to use async sign_in

jacksontj committed:
We aren't doing singleton module loaders, lets not clear what we just made

jacksontj committed:
Remove threading from MultiMinion

jacksontj committed:
Cleanup

jacksontj committed:
More cleanup

jacksontj committed:
Make destroy properly unload things

jacksontj committed:
Remove threading from MultiSyndic

jacksontj committed:
Fix syndic event forwarding

jacksontj committed:
Change pubchannel over in zeromq

jacksontj committed:
Remove remnants of old Sync pub channel

jacksontj committed:
Add basic Asyncpubchannel tests to zmq

jacksontj committed:
Add unregister tests

jacksontj committed:
Refactor pub tests

jacksontj committed:
Fix misc tests

jacksontj committed:
Cleanup shutdown of tests

jacksontj committed:
Always attempt to send errors to clients

jacksontj committed:
Fix mocking, not quite sure how this magic mock works

jacksontj committed:
Merge cleanup

jacksontj committed:
Fix rebase collisions

there were some refactors which have happened over the last month, fixing those

jacksontj committed:
Misc cleanup

jacksontj committed:
Make Sync wrapper for async req channels

jacksontj committed:
Remove sleep from saltmessageclient

Now we will only run the send routine if we have something to send out. This reduces CPU load (no unecessary runs) and latency (no sleep)

jacksontj committed:
Switch tcp req channel over to async auth libraries

jacksontj committed:
Reset asyncauth singleton in tests

jacksontj committed:
Make AsyncAuth singleton per io_loop

In addition this makes the singleton instance mapping use weakrefs-- so if the io_loop goes away (or we don't reference the object anywhere) we will remove the auth instance. This allows for us to "unload" an auth without having to exit the interpreter or forcibly delete it

jacksontj committed:
Move syncwrapper to utils-- since we'll have more uses for it

jacksontj committed:
Change the Auth base class to AsyncAuth

jacksontj committed:
Add weakref to SAuth

jacksontj committed:
Fix exceptional case where no timeout is set

jacksontj committed:
Properly bubble up auth failures to the future

jacksontj committed:
Attempt to return nice errors even if it wasn't properly msgpackd

jacksontj committed:
Migrate zeromq req channel to async interfaces

jacksontj committed:
Fix zmq pub channels

jacksontj committed:
Better message_id handling

This caps the number of inflight messages to maxint on the system, so we can wrap the message IDs instead of forever increasing in size

jacksontj committed:
Cleanup auth future, and pass io_loop to instance

jacksontj committed:
Overwrite the current IOLoop for blocking calls

This adds a context manager which will temporarily overwrite the current ioloop for these blocking calls. This way you can safely call blocking methods which are made on top of ioloops without library methods attaching your future to the wrong ioloop.

jacksontj committed:
Change _handle_payload to return a future.

This means we can make more and more of the core async (using coroutines) without having to change the transport modules-- as they all already expect futures.

jacksontj committed:
Share creds between all channels (if you can).

This allows sync and async callers to Auth to have access to the same creds.

jacksontj committed:
Fix calls to send() to not specify a callback of int

jacksontj committed:
Handle case where we miss first sign-in

jacksontj committed:
Make transport unit tests pull configs from test suite instead of from /etc/salt

jacksontj committed:
SNDTIMEO is not implemented in pyzmq, and since we are using linger this gets us the same thing

jacksontj committed:
Term context of AsyncEventPublisher on destroy()

jacksontj committed:
Make sure to hand out RAET channels on the old interface

jacksontj committed:
Simplify TCP wire protocol

jacksontj committed:
More cleanup of tcp wire framing-- pushing msgpacks down a layer

jacksontj committed:
Lint cleanup

jacksontj committed:
Fix merge

jacksontj committed:
Add tornado dependancy

jacksontj committed:
Fix race condition in integration test startup

jacksontj committed:
Remove "close()" api-- just use a destructor

jacksontj committed:
Cleanup AsyncEvent publisher

And remove PollingEventPublisher

jacksontj committed:
Add basic tests for asyncEventpublisher

jacksontj committed:
Make sure to close server processes after each test.

Waiting for the destructor doesn't necessarily work, since its not gauranteed to run immediately.

jacksontj committed:
Mock/stub out _send_pub in ClearFuncs

Before we were attempting to serialize a MagicMock object

s0undt3ch commented:
I just woke up and had my demanding coffee, which means that I don't yet have enough active brain cells to read the code but you got my curiosity pumped up. :smile:

jacksontj committed:
Pylint

jacksontj committed:
Pylint

jacksontj committed:
Pylint cleanup in TCP transport module

jacksontj committed:
Pylint

jacksontj committed:
pylint cleanup

jacksontj committed:
pylint cleanup

jacksontj committed:
pylint cleanup

jacksontj committed:
pylint cleanup

jacksontj committed:
pylint cleanup

jacksontj committed:
pylint cleanup

jacksontj committed:
pylint cleanup

jacksontj committed:
pylint cleanup

jacksontj committed:
pylint cleanup

jacksontj committed:
Clean up flo zmq deed, since we have changed the MWorker

jacksontj committed:
pylint cleanup

jacksontj committed:
pylint cleanup

jacksontj committed:
pylint cleanup

jacksontj committed:
Remove AES key initing-- since we are mocking out _send_pub

jacksontj committed:
pylint cleanup

jacksontj committed:
pylint cleanup

mgwilliams commented:
Echoing @jfindlay's "Awesome"ness... especially the async part.

jacksontj committed:
Correct comment of wire-protocol

jacksontj committed:
Add initial transport docs

jacksontj commented:
I just added the first cut of docs (3b7170d), please let me know of specific questions you have and I can incorporate them into the docs.

cachedout commented on doc/topics/transports/index.rst:
```diff
@@ -0,0 +1,35 @@
+primarily used for fetching files and returning job returns. The req channels
+have two basic interfaces when talking to the master. ``send`` is the basic
+method that gaurantees the message is encrypted at least so that only minions
+attached to the same master can read it-- but no gaurantee of minion-master
```
```guarantee```

jacksontj replied:
Thanks! I always misspell that :(

jacksontj committed:
typo fixing!

jacksontj committed:
pylint cleanup

jacksontj committed:
pylint cleanup

jacksontj committed:
pylint cleanup

jacksontj commented:
For some reason I'm getting pylint errors for importing tornado.gen?? (https://jenkins.saltstack.com/job/salt-pr-lint-n/4004/violations/file/salt/master.py/) Any ideas?

s0undt3ch commented:
Is `tornado.gen` a "weird" module? It could just be pylint not knowing about that "weird" module. Pylint fails on `distutils.version`(as an example)...

cachedout commented on salt/minion.py:
```diff
@@ -1415,75 +1233,6 @@ def _refresh_grains_watcher(self, refresh_interval_in_minutes):
-                zmq.TCP_KEEPALIVE_INTVL, self.opts['tcp_keepalive_intvl']
-            )
-
-    def _set_monitor_socket(self):
```
Did this functionality get re-implemented? I'm still finding my way around these changes but I don't see it.

jacksontj replied:
Already done-- it just moves into the pubchannel (https://github.com/saltstack/salt/pull/22453/files#diff-7e93fb0d8129a1b378ae1a10559936cbR165)

cachedout replied:
Forgive me if I'm missing something, but I was referring to the monitor sockets, not the keepalives.

cachedout commented:
Finished my first pass on reading this. I think this is a really sound approach, with many extremely substantial improvements!

I'll continue to digest this as I find the time to really dig in here. The first concern that comes to mind here is Windows support, though. To the best of my knowledge, Tornado is not considered production-ready on Windows, so we'll need to really test hard in that area. @jacksontj any thoughts there?

Of course, I think we're a ways off from actually getting this merged but I'm thrilled to see this work. It's really wonderful.

I'm looking forward to actually playing around with this code this afternoon and seeing what it can do.

thatch45 commented:
My main concern here is windows and mac, as the tornado docs specify that they are for development only at this point

thatch45 commented:
Also, does tornado hard dep certifi? Some disttros don't like certifi

thatch45 commented:
I guess what I am saying is that I don't like the tornado hard dep addition, but I do like the idea of being able to eventually move away from ZMQ

mgwilliams commented:
@thatch45 The docs say OS X is recommended for development due to performance issues. Some Googling seems to indicate the Windows issue is with `tornado.autoreload`, which is slated to be fixed in the next release. Probably performance issues on Windows too ;-).

thatch45 commented:
It makes me wonder if we could add this without making tornado a hard dep, I think that this needs a lot more testing before we throw it in like this. Also, we JUST got the master running on windows, I don't want to see that damaged

thatch45 commented:
We also need to verify all tornado deps, make sure we don't have anything nasty in there....

jacksontj commented:
@s0undt3ch What do you mean by "weird"? AFAIK its just a regular module..?

jacksontj commented:
@thatch45 yea, we'll definitely need to look over all the deps, but this does seem like the way forward. As @mgwilliams said the problems on Mac OS + Windows are performance related-- so we should be able to fix them. But from the reading I've done it seems to be tied with tornado's webserving etc. not with the concurrency stuff (futures, coroutines, etc.) which is what we primarily use.

We chatted at SaltConf and you didn't bring up concerns about adding the tornado dep-- but if you have concerns we can talk about them, what concerns (excluding the perf stuff) do you have?

thatch45 commented:
That sounds good on the macos and windows stuff, it looks like we still need to remove zmq from the ipc systems, raw stream sockets might be a good approach. I will look over the deps to make sure we don't have any issues in there.
Also, if we can get rid of the M2Crypto dep it would help us justify this and work towards a move to asyncio and python3. I think we should establish a going forward plan to merge all of this stuff in a larger effort.

jacksontj commented:
From my perspective this is just step 1 of the overhaul. This gets you abstracted transports (so you can remove zmq) and a way to change up the transport in a modular way. Once this is in I plan on making zeromq/tcp (the auth mixin) support upgrading to nacl-- then after that we could remove the dep for M2crypto. IMO we shouldn't block this merging on removing M2crypto-- primarily because this is going to become a hard feature branch to keep in sync ;)

jacksontj commented:
Jenkins seems to be running out of FDs, maybe I have a leak? I didn't see anything locally.. I'll run them again locally though to double check.

jacksontj committed:
fix typo

thatch45 commented:
I agree, I have reviewed the tornado deps and I am feeling good about them, there are no hard deps or optional deps. This is a large change so I am hesitant as I am sure you can imagine :), but we want to get it in.
I agree on the sockets, we should support local tcp sockets for windows and stream unix sockets as they are faster (Or so is my current understanding, they are certainly faster in ZMQ).
I am still trying to figure out the certifi relationship of tornado although this use case would not mandate it.

Also, if we can make that natural move to libnacl we should be able to support a full python stack, so I am all on board there as well.

s0undt3ch commented:
The "weird" module concern is that some libraries create modules at runtime and of course, pylint has issues with those... would be great if it hadn't... I'll have to checkout this code and run pylint myself to see what's it's problem...

jacksontj commented:
Completely agree on the IPC stuff :)

As for libnacl-- yea, once we add support for that and the migration we can make M2Crypto an optional dep (for backwards compat), but we could drop it moving forward. I have the migration plan written down in some notebook of mine, basically we'd do an NPN style negotiation on sign-in that would determine the crypto. I was also considering trying out using openssl to just wrap the sockets, but that might be for another day ;)

thatch45 commented:
Ok, once we have the tests running I am giving this the thumbs up, then we can take care of the most pressing matter which is making the tornado transport not dep on ZMQ for IPC and then we should be in a stable place

thatch45 commented:
Oh, we should also figure out a "revert" plan, just in case. We have traveled down a few roads in the past that we have wanted to later back out of and we should be careful on this one, this is a big change :)

thatch45 commented:
Oh, and you might want to read up on curvecp:
http://curvecp.org/

cachedout commented on salt/transport/zeromq.py:
```diff
@@ -0,0 +1,636 @@
+
+# tornado imports
+import tornado
+import tornado.tcpserver
```
Are the tornado.tcpserver and tornado.tcpclient imports needed in this module? I don't see them used.

jacksontj replied:
They aren't ;) wonder why pylint didn't call me out on that :/

thatch45 commented:
I am also going to ask @twangboy to make sure this will still run on the minion in windows

jacksontj commented:
@thatch45 Yea, revert plan would be to revert all these commits ;) As for the tests-- they just all passed locally... so is jenkins having problems? Or just me?

jacksontj committed:
Remove unused imports

cachedout commented:
I have to say, after spending the day with this code, I *really* love it. It takes some getting used to but this approach is absolutely brilliant.

thatch45 commented:
Jenkins does not run in a very beefy environment, on purpose, I wonder if that might be it?

jacksontj commented:
possible.. do you know what the ulimits are? (or can you paste me the ulimit -a as the user on jenkins)? I run the tests locally as root so its possible I'm doing some lazy closing of sockets which allows too many to pile up during the test run ;)

jacksontj commented:
@cachedout Yea, the futures stuff is awesome. I'm glad that's what python is doing moving forward. Its a great way of thinking about concurrency in a way that doesn't twist your mind ;)

cachedout commented:
When running the tests locally as root on my machine:

```
FAILED (errors=42, skipped=72)
Traceback (most recent call last):
  File "./runtests.py", line 466, in <module>
    main()
  File "./runtests.py", line 451, in main
    status = parser.run_integration_tests()
  File "./runtests.py", line 385, in run_integration_tests
    status.append(self.run_integration_suite('cli', 'CLI'))
  File "./runtests.py", line 244, in run_integration_suite
    return self.run_suite(path, display_name)
  File "/home/mp/devel/salt-testing/salttesting/parser/__init__.py", line 481, in run_suite
    tests = loader.discover(path, suffix, self.testsuite_directory)
  File "/usr/lib/python2.7/unittest/loader.py", line 206, in discover
    tests = list(self._find_tests(start_dir, pattern))
  File "/usr/lib/python2.7/unittest/loader.py", line 241, in _find_tests
    paths = os.listdir(start_dir)
OSError: [Errno 24] Too many open files: '/home/mp/devel/salt/tests/integration/cli'
```

jacksontj commented:
@cachedout Hmm, just running "python runtests" ?

cachedout commented:
sudo python2 ./runtests.py

cachedout commented:
It runs all right until about midway through the module tests, then it starts to tank. Once, I saw it on the test runner output but now I'm having to watch /tmp/salt-runtests.log. By midway through the module tests, stacktraces will start flying by.

jacksontj commented:
Hmm, I'll start the run up again locally... :/

twangboy commented:
All my tests work on windows.

thatch45 commented:
W000t! @twangboy !!

jacksontj commented:
@cachedout Yea, I'm still unable to reproduce this locally :( What OS are you running? I tried on fedora 20 and rhel 6

cachedout commented:
@jacksontj It's Arch. 
```
mp@silver ...devel/salt/tests % salt --versions-report                                              (git)-[remotes/tj/transport] 
Salt Version:
           Salt: 2015.2.0-1127-g106f091
 
Dependency Versions:
         Jinja2: 2.7.3
       M2Crypto: 0.22
           Mako: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.5.0
         Python: 2.7.9 (default, Dec 11 2014, 04:42:00)
           RAET: 0.6.1
            ZMQ: 4.0.5
          ioflo: 1.1.9
        libnacl: 1.4.0
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
       pycrypto: 2.6.1
 
System Versions:
           dist:   
        machine: x86_64
        release: 3.19.3-1-ARCH
```

UtahDave commented:
@thatch will this change or improve on the threading situation on Windows?

jacksontj commented:
@UtahDave It should more or less fix it-- since this is using coroutines (basically stack-less threads). Assuming the issue with threading on windows was that threads are weird ;) Tornado makes threading in Salt unnecessary.

jacksontj commented:
@cachedout Hmm, I'm trying to run the tests with the arch docker container (--docked=arch), we'll see.

jacksontj commented:
@cachedout Yea, --docked doesn't work:

```
python runtests.py --docked=arch
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Docker command: python2 /salt-source/tests/runtests.py --run-destructive
 * Running the tests suite under the 'salttest/arch' docker container. CID:Warning: '-cidfile' is deprecated, it will be replaced by '--cidfile' soon. See usage.
 586f0c51b4881689006fa7d9df27e3f9a72992fd8c5ba19ca92e2162715d8671
                                                                 ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Running: supervisord
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
/usr/lib/python2.7/site-packages/supervisor/options.py:296: UserWarning: Supervisord is running as root and it is searching for its configuration file in default locations (including its current working directory); you probably want to specify a "-c" argument specifying an absolute path to a configuration file for improved security.
  'Supervisord is running as root and it is searching '

---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Running: python2 /salt-source/tests/runtests.py --run-destructive
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
python2: can't open file '/salt-source/tests/runtests.py': [Errno 2] No such file or directory
```

cachedout commented:
@jacksontj Sounds good. I'll also see if I can reproduce this elsewhere and if I can, I'll shoot you over some credentials to login and have a look.

cachedout commented:
@jacksontj Oh, I don't think we've used docked in a while. That was something @s0undt3ch was playing around with a while back but he may have abandonded it.

jacksontj commented:
If so we should probably rip it out ;)

cachedout commented:
Agreed on the docker flag. :]

I have a VM up now that's exhibiting the problem with the test suite. I will shoot over details to you in an email.

jacksontj commented:
@cachedout Awesome! I'll take a look at it either tonight or tomorrow morning ;)

s0undt3ch commented:
We try to raise ulimit to a max of [3072](https://github.com/saltstack/salt/blob/develop/tests/runtests.py#L40), [but only in case the current value is lower than that](https://github.com/saltstack/salt/blob/develop/tests/runtests.py#L294-L318).

As to the `--docked` flag, I feel like it could be stripped out(needs some updating) but when working I find it useful. Eventually,when I finish and merge my [salt-runtests branch](https://github.com/s0undt3ch/salt/tree/features/salt-runtests) which is now 10698 commits behind :frowning:, I could pick up, or also drop docker support. The current status is that docker has evolved and we need to maintain some docker images and time right now is not that much(but it's not impossible).

@cachedout it ain't abandoned, it just needs some serious lovin' :smile: 

@jacksontj I can prep you an Arch(or whichever you prefer) vm, just like those we run on Jenkins and give you access to it(like we did before)...

s0undt3ch commented:
Oh, @cachedout beat me to that VM :smile: 

I should have read all comments before posting...

cachedout commented:
@s0undt3ch Yes, I didn't know if there were plans to revive or not. I was just saying it wasn't currently in use anywhere, so we could have temporarily removed the flag or put a note in the help docs.

I have tried on my local machine to run with a ulimit for nofiles of 4000 and while it did delay the error somewhat, it still occurred.

s0undt3ch commented on salt/transport/local.py:
```diff
@@ -0,0 +1,41 @@
+            #data = json.loads(load)
+            #{'path': 'apt-cacher-ng/map.jinja', 'saltenv': 'base', 'cmd': '_serve_file', 'loc': 0}
+            #f = open(data['path'])
+            f = open(load['path'])
```
`salt.utils.fopen()`

s0undt3ch replied:
Oh, testing purposes.... well, doesn't hurt.

jacksontj replied:
Yea, I can change that over.. I didn't actually write this one, I just moved it ;) I'm not sure why its needed as you can easily test the channels without using this one... I think its legacy so I'd like to remove it at some point ;)

jacksontj committed:
More proactive cleanup in sync wrapper

Before we were relying on the underlying objects to not reference the ioloop, now on destruction we'll specifically close it, freeing its internal FDs

jacksontj committed:
Destructors for the transport classes

Since these are using IOLoops, the underlying transport class (message_client) still leaves FDs attached to the IOLoop. This means we have to specifically close them out when we destroy our class, otherwise the ioloop still references them, and python won't GC them.

jacksontj commented:
@cachedout 82990c0 should be our fix. Basically when we create the sockets it registers them with the IOloop, if we destroy our class without unregistering it the ioloop maintains a reference-- and won't close it. So that meant that in the tests (where we create and destroy many a class) we were piling up tons. This passes on the VM you provided, so I expect it to pass just fine in jenkins :)

jacksontj commented:
BTW assuming the tests pass fine we can tear down the VM ;)

cachedout commented:
Great! I will start testing this here as well. Nice fix!

cachedout commented:
Ran perfectly for me!

cachedout commented:
Jenkins seems to be throwing these:

```
00:08:50.402 Exception TypeError: 'heap argument must be a list' in <bound method AsyncZeroMQReqChannel.__del__ of <salt.transport.zeromq.AsyncZeroMQReqChannel object at 0x7f93b4f88950>> ignored
00:08:50.403 Exception TypeError: 'heap argument must be a list' in <bound method AsyncZeroMQReqChannel.__del__ of <salt.transport.zeromq.AsyncZeroMQReqChannel object at 0x7f93d96288d0>> ignored
```

cachedout commented:
Oddly, Jenkins has died with the same error:
```
00:14:05.789 Too many open files (epoll.cpp:39)
```

jacksontj commented:
There are a few similar cases where I'm not cleaning up the FDs yet, I'm working on adding a check for the leaks in the unit tests before I commit it. It is odd though that jenkins ran into the issue but not the VM, although I guess the VM probably has more available than jenkins.

jacksontj committed:
Make sure to close sockets on deletion of pubchannels as well

jacksontj committed:
Check for FDs attached to the IOLoops after test runs

jacksontj committed:
Add workaround to a bug in pyzmq's stream.close()

jacksontj commented:
These should help a bunch. Depending on version of pyzmq there are bugs in how it closes streams, specifically for me (14.0.1) it will throw that heap argument must be a list thing.... I added the workaround code for it which we could remove if we forced people to newer ones, but the workaround code is quite simple so we may just keep it.

cachedout commented:
Go Go Jenkins!

cachedout commented:
@jacksontj Great! We're all on board with getting this merged once the test suite give the thumbs up. Let's see how this run goes. :]

jacksontj committed:
pylint

jacksontj committed:
Set a linger of 0 on the stream's socket, and dont try to close twice

cachedout commented:
Go Go Jenkins!

cachedout commented:
Jenkins seems to be having a tough time today.

s0undt3ch commented:
Only the lint error in salt.crypt should be considered, the others can be ignored.

So, its close!

jacksontj commented:
I'll grab those, one sec.

jacksontj commented:
Is there a way to run pylint locally? I tried it before and I remember many many exceptions..

jacksontj committed:
pylint cleanup

s0undt3ch commented:
Install PyLint and SaltPyLint and you can run it with salt's `.testing.pylintrc`

jacksontj commented:
These pylint failures are unrelated to my changes... I'll rebase to see if they go away.

jacksontj commented:
Apparently this is against the newest develop, did someone make a change to pylint?

s0undt3ch commented:
I installed tornado to get rid of the import failure :)

s0undt3ch commented:
Non of the lint failures are related to this change.

However, if you find some free time, since saltnado is also your work, could you clean those up? probably best in a different PR.

jacksontj commented:
@s0undt3ch #22555 should be our lint fix, I can pull them into this branch to clear the lint failures if we want, but its probably not necessary ;)

cachedout commented:
OK, buckle up. Here we go. :]

s0undt3ch commented:
WooHooo!

whiteinge commented:
_straps in and throws his hands in the air_

cachedout commented:
@jacksontj Just an FYI. We're seeing some bumpiness now that we're testing across multiple platforms. Below are the current issues. I'll try to keep this updated as best I can and I'll start looking into these as well:

https://jenkins.saltstack.com/job/salt-linode-ubuntu14.04/1404/testReport/unit.transport.zeromq_test/ClearReqTestCases/test_basic/

https://jenkins.saltstack.com/job/salt-linode-ubuntu14.10/97/testReport/unit.utils.event_test/TestSaltEvent/test_event_nested_subs/

jacksontj commented:
Hmm, odd that we didn't see that in jenkins yea? If you guys need help looking into stuff I should have some time today :)

cachedout commented:
Yeah, odd that we didn't see that before. I have no explanation for that. :]

If you have a minute and can look, it's always very appreciated. :]

jacksontj commented:
@cachedout Are those 2 easily reproducible on Ubuntu? I ran stuff again locally and I'm unable to repro locally... If so maybe time for another VM ;)

jacksontj commented:
Also, is there a way to see log output from those test failures?

cachedout commented:
Yes. The build artficats for the two errors listed above are here, respective to the order listed originally:

ClearReqTest:
https://jenkins.saltstack.com/job/salt-linode-ubuntu14.04/1404/

NestedSubTest
https://jenkins.saltstack.com/job/salt-linode-ubuntu14.10/97/

You can also use 'Console View' on the left of each of those to get results from the test runner itself. If that's not enough, I can spin up a VM and see if I can get these errors to re-occur. I haven't seen them before this, though, so perhaps they are transient? We'll see. :]

jacksontj commented:
Hmm, for both of there there is absolutely nothing in the logs:

```
17:36:12,705 [salttesting.xmlunit                                 :76  ][DEBUG   ] >>>>> START >>>>> unit.transport.zeromq_test.ClearReqTestCases.test_basic
17:36:14,733 [salttesting.xmlunit                                 :90  ][DEBUG   ] <<<<< END <<<<<<< unit.transport.zeromq_test.ClearReqTestCases.test_basic
```


```
17:16:10,845 [salttesting.xmlunit                                 :90  ][DEBUG   ] <<<<< END <<<<<<< unit.utils.event_test.TestSaltEvent.test_event_nested_subs
17:16:10,846 [salttesting.xmlunit                                 :76  ][DEBUG   ] >>>>> START >>>>> unit.utils.event_test.TestSaltEvent.test_event_no_timeout
```

I wonder if we just provoked some race condition (zeromq has many..)

cachedout commented:
That would not surprise me even a little bit. :] The thing to do here may be to wait until we get a little more information from the test suite and see which tests continue to be unstable and then try to come up with an easier-to-reproduce test case.

jacksontj commented:
Yea, that might be good. Especially if its specific platforms which show the failures all the time-- we have found weird bugs in the python bindings before which only manifest in some distros (due to versioning).

cachedout commented:
Indeed. I'll keep an eye on it and start to look for patterns that might be useful.

s0undt3ch commented:
There's also salt-runtests.log in the build artifacts if Jenkins was able to download it.

cachedout commented:
@s0undt3ch I think that's what he pasted the logs from, actually.

s0undt3ch commented:
Oh! :)

cachedout commented:
Looks like we're having problems on Ubuntu 10 and Ubuntu 12. Seems to be related to pyzmq 13.x.

https://gist.github.com/rallytime/e8c08f3de470b17ee8c1

We're looking into it as well, but just a heads up.

jacksontj commented:
@rallytime @cachedout Looks like the name for ZMQIOloop (in the pyzmq bindings) came about in v14.0.0, if we need to support v13.0.0 we'll have to make some helper function to conditionally import it with the old name. IMO we should just upgrade it ;)

s0undt3ch commented:
We should still support v13.x... Not every user will be willing to update pyzmq unless its packaged, even though we still have some time before we release this code in a major release.

We already advise users to upgrade to v14.x but so far we're not forcing anyone...

jacksontj commented:
Ok, then we'll need to conditionally import... I can throw together a PR in just a minute ;)

s0undt3ch commented on salt/transport/zeromq.py:
```diff
@@ -0,0 +1,654 @@
+        raise tornado.gen.Return(ret)
+
+    @property
+    def stream(self):
```
Any specific reason why we don't init the `self.stream` reference on the class'es `__init__` method?

The reason I ask this is because while writing a `py.test` plugin which should get a salt master and minion running, and because my current test is just to make sure they are up, ie, no interaction besides connecting has happened before the daemons are shutdown, I'm getting the following error:
```
Exception TypeError: "'NoneType' object is not callable" in <bound method Minion.__del__ of <salt.minion.Minion object at 0x7fa5747fa710>> ignored
```

Somewhere on the teardown code of, I think zmq, `zmq.eventloop.zmqstream` becomes `None`, and, at that stage, in this stream method, we can't even import `zmq.eventloop.zmqstream`.

However, if I instantiate the stream on class initialization, I no longer get the error.

Before submitting a pull request I just wanted to know why the deferred initialization on the zmqstream...

cachedout replied:
I do not know the specific reason for this, nor can I think of one offhand
but @jacksontj would be the one to provide the best insight here.

On Tue, Oct 20, 2015 at 2:04 AM, Pedro Algarvio <notifications@github.com>
wrote:

> In salt/transport/zeromq.py
> <https://github.com/saltstack/salt/pull/22453#discussion_r42465926>:
>
> > +        '''
> > +        messages_len = len(messages)
> > +        # if it was one message, then its old style
> > +        if messages_len == 1:
> > +            payload = self.serial.loads(messages[0])
> > +        # 2 includes a header which says who should do it
> > +        elif messages_len == 2:
> > +            payload = self.serial.loads(messages[1])
> > +        else:
> > +            raise Exception(('Invalid number of messages ({0}) in zeromq pub'
> > +                             'message from master').format(len(messages_len)))
> > +        ret = yield self._decode_payload(payload)
> > +        raise tornado.gen.Return(ret)
> > +
> > +    @property
> > +    def stream(self):
>
> Any specific reason why we don't init the self.stream reference on the
> class'es __init__ method?
>
> The reason I ask this is because while writing a py.test plugin which
> should get a salt master and minion running, and because my current test is
> just to make sure they are up, ie, no interaction besides connecting has
> happened before the daemons are shutdown, I'm getting the following error:
>
> Exception TypeError: "'NoneType' object is not callable" in <bound method Minion.__del__ of <salt.minion.Minion object at 0x7fa5747fa710>> ignored
>
> Somewhere on the teardown code of, I think zmq, zmq.eventloop.zmqstream
> becomes None, and, at that stage, in this stream method, we can't even
> import zmq.eventloop.zmqstream.
>
> However, if I instantiate the stream on class initialization, I no longer
> get the error.
>
> Before submitting a pull request I just wanted to know why the deferred
> initialization on the zmqstream...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/saltstack/salt/pull/22453/files#r42465926>.
>

s0undt3ch replied:
A simple test.ping will create the stream instance, so, for now, I've sidestepped the issue but my question is still valid.

jacksontj replied:
This is just simply lazy creation of the socket. This is really just a holdover of what was done before (we didn't make it until it was needed). We can change it up, but we'd have to make sure that all clients of this class are okay with the socket being inited when the class is (since zmq isn't process/thread safe).

s0undt3ch replied:
My problem is very particular and seems unrelated to the regular workflow, so, lets leave it alone, at least while an issue regarding regular salt workflow comes up. I can just issue a test.ping(which I actually do need to) for what I'm currently working on...
